src/shared_memory_manager.h

```diff
@@ -72,6 +72,7 @@ class SharedMemoryManager {
     void* mapped_addr_;
     TRITONSERVER_MemoryType kind_;
     int64_t device_id_;
+    bool awaiting_unregister_;
```

Comment:
question: could this be done with smart pointer referencing counting instead of an explicit flag?

Reply:
@nnshah1, yes, I believe we can implement this with `weak_ptr`.
@GuanLuo has previously provided input on an alternative approach using `weak_ptr` and also discussed the current method.
Here’s a summary of the `weak_ptr` approach:
Currently, the shared memory (shm) manager uses a map of shared_ptr for shm_info. When the unregister API is called on a region still in use, we can convert it to a `weak_ptr` to decrease the reference count. We would track these in a separate `weak_ptr map` until fully unregistered, preventing new regions with the same name and allowing status check.

When the last reference to the `shm_info` `shared_ptr` is released (after inference completes), the `shared_ptr` destructor will trigger. At that point, we can invoke the unregister logic and remove the associated `weak_ptr` from the second map.

To support this approach, we would likely need to modify most of shm manager methods, such as register, status, unregister, and unregisterHelper, etc. to work with the `weak_ptr map`.

For simplicity, I proceeded with the current approach. Please let me know if we’d like to explore the `weak_ptr` approach, and I can work on it. Happy to connect and discuss if any optimizations are needed.
Thank you.

Reply:
Thanks for the detailed response - it was just that it seemed like a lot of similarity with reference counting and shared pointers - so was hoping it was something quick. But if this is the more practical short term - no issue, 

@GuanLuo - would we want to pre-emptively put a refactor story in? (only if you think its a near term thing that needs to be addressed) -

Reply:
I have raised the point that using smart pointer is a cleaner way to address this problem early on for this development, so you can have a refactor story as it is just a right thing to do. However, I believe that it is just going to be another form of TODOs at this point and what is in this PR will remain unchanged for a long time.